pkg/sources/github/feedlet.go

```diff
@@ -0,0 +1,452 @@
+		feedNamespace:          feedNamespace,
+		feedServiceAccountName: feedServiceAccountName,
+		image: image,
+	}
+}
+
+// TODO(n3wscott): Add a timeout for StopFeed.
+func (t *githubEventSource) StopFeed(trigger sources.EventTrigger, feedContext sources.FeedContext) error {
+	log.Printf("stopping github webhook feed with context %+v", feedContext)
+
+	return t.deleteWebhook(trigger, feedContext)
+}
+
+// TODO(n3wscott): Add a timeout for StartFeed.
+func (t *githubEventSource) StartFeed(trigger sources.EventTrigger, target string) (*sources.FeedContext, error) {
+
+	// Create the Receive Adapter Service that will accept incoming requests from GitHub.
+	service, err := t.createReceiveAdapter(trigger, target)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create service: %v", err)
+	}
+
+	// TODO(n3wscott): look into using spew.
+	log.Printf("created Service: %+v", service)
+
+	// Start watching the Receive Adapter Service for it's updated domain name. This will be passed
+	// to GitHub as part of the webhook registration.
+	receiveAdapterDomain, err := t.waitForServiceDomain(service.GetObjectMeta().GetName())
+	if err != nil {
+		return nil, fmt.Errorf("failed to get the service: %v", err)
+	}
+
+	return t.createWebhook(trigger, service.GetObjectMeta().GetName(), receiveAdapterDomain)
+}
+
+func (t *githubEventSource) waitForServiceDomain(serviceName string) (string, error) {
+	sc := t.servingclientset.ServingV1alpha1().Services(t.feedNamespace)
+
+	wt := int64(watchTimeout / time.Second)
+	opts := metav1.ListOptions{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "Service",
+			APIVersion: "v1alpha1",
+		},
+		FieldSelector:  fmt.Sprintf("metadata.name=%s", serviceName),
+		LabelSelector:  "receive-adapter=github",
+		Watch:          true,
+		TimeoutSeconds: &wt,
+	}
+
+	w, err := sc.Watch(opts)
+	if err != nil {
+		log.Printf("[%s] failed to create watch: %v", serviceName, err)
+		return "", err
+	}
+
+	eventChan := w.ResultChan()
+	for {
+		event, more := <-eventChan
+		if !more {
+			return "", fmt.Errorf("[%s] watch channel closed, no more", serviceName)
+		}
+		switch event.Type {
+		case watch.Error:
+			return "", fmt.Errorf("[%s] watched Service error", serviceName)
+		case watch.Deleted:
+			return "", fmt.Errorf("[%s] watched Service deleted", serviceName)
+		case watch.Added, watch.Modified:
+			service, ok := event.Object.(*v1alpha1.Service)
+			if !ok {
+				log.Printf("[%s] expected a Service object, but got %T", serviceName, event.Object)
+				continue
+			}
+			if service.Name != serviceName {
+				log.Printf("Error: [%s] expected a service.Name %q to match expected serviceName %q",
+					serviceName, service.Name, serviceName)
+				continue
+			}
+			status := service.Status
+			if status.Domain != "" {
+				w.Stop()
+				return status.Domain, nil
+			}
+
+		}
+	}
+}
+
+func receiveAdapterName(trigger sources.EventTrigger) string {
+	// TODO(n3wscott): this needs more UUID on the end of it?
```

Comment:
Can you capture the issue that we don't take the event type into account so different event types will end up with the same name and clobber each other?